pkg/plugins/policies/meshretry/api/v1alpha1/validator.go

```diff
@@ -0,0 +1,166 @@
+		SupportedKinds: []common_api.TargetRefKind{
+			common_api.Mesh,
+			common_api.MeshSubset,
+			common_api.MeshService,
+			common_api.MeshServiceSubset,
+		},
+	})
+	return targetRefErr
+}
+
+func validateTo(to []To) validators.ValidationError {
+	var verr validators.ValidationError
+	for idx, toItem := range to {
+		path := validators.RootedAt("to").Index(idx)
+		verr.AddErrorAt(path.Field("targetRef"), matcher_validators.ValidateTargetRef(toItem.TargetRef, &matcher_validators.ValidateTargetRefOpts{
+			SupportedKinds: []common_api.TargetRefKind{
+				common_api.Mesh,
+				common_api.MeshService,
+			},
+		}))
+		verr.AddErrorAt(path.Field("default"), validateDefault(toItem.Default))
+	}
+	return verr
+}
+
+func validateDefault(conf Conf) validators.ValidationError {
+	var verr validators.ValidationError
+	path := validators.RootedAt("conf")
+	if conf.TCP == nil && conf.HTTP == nil && conf.GRPC == nil {
+		verr.AddViolationAt(path, "at least one of the 'tcp', 'http', 'grpc' must be specified")
+	}
+	if conf.TCP != nil {
+		verr.AddErrorAt(path, validateTCP(conf.TCP))
+	}
+	if conf.HTTP != nil {
+		verr.AddErrorAt(path, validateHTTP(conf.HTTP))
+	}
+	if conf.GRPC != nil {
+		verr.AddErrorAt(path, validateGRPC(conf.GRPC))
+	}
+	return verr
+}
+
+func validateTCP(tcp *TCP) validators.ValidationError {
+	var verr validators.ValidationError
+	path := validators.RootedAt("tcp")
+	if tcp.MaxConnectAttempt == nil {
+		verr.AddViolationAt(path, validators.MustNotBeEmpty)
+	}
+	return verr
+}
+
+func validateHTTP(http *HTTP) validators.ValidationError {
+	var verr validators.ValidationError
+	path := validators.RootedAt("http")
+	if http.NumRetries == nil && http.PerTryTimeout == nil && http.BackOff == nil && http.RetryOn == nil &&
+		http.RetriableRequestHeaders == nil && http.RetriableResponseHeaders == nil {
+		verr.AddViolationAt(path, validators.MustNotBeEmpty)
+	}
+	if http.BackOff != nil {
+		verr.AddErrorAt(path, validateBackOff(http.BackOff))
+	}
+	if http.RetryOn != nil {
+		verr.AddErrorAt(path, validateHTTPRetryOn(http.RetryOn))
+	}
+	return verr
+}
+
+func validateHTTPRetryOn(retryOn []HTTPRetryOn) validators.ValidationError {
+	var verr validators.ValidationError
+	path := validators.RootedAt("retryOn")
+	for idx, ro := range retryOn {
+		switch ro {
+		case ALL_5XX:
+		case GATEWAY_ERROR:
+		case RESET:
+		case RETRIABLE_4XX:
+		case CONNECT_FAILURE:
+		case ENVOY_RATELIMITED:
+		case REFUSED_STREAM:
+		case HTTP3_POST_CONNECT_FAILURE:
+		case HTTP_METHOD_CONNECT:
+		case HTTP_METHOD_DELETE:
+		case HTTP_METHOD_GET:
+		case HTTP_METHOD_HEAD:
+		case HTTP_METHOD_OPTIONS:
+		case HTTP_METHOD_PATCH:
+		case HTTP_METHOD_POST:
+		case HTTP_METHOD_PUT:
+		case HTTP_METHOD_TRACE:
+		default:
+			// method http.StatusText returns empty string for unknown status codes
+			if i, err := strconv.Atoi(string(ro)); err != nil || http.StatusText(i) == "" {
```

Comment:
oh, I didn't know I can put status code like `422`. Can we add this to API Schema desc?